flake8_trio.py

```diff
@@ -55,19 +83,73 @@ def __init__(self) -> None:
         self.safe_yields: Set[ast.Yield] = set()
         self._yield_is_error = False
         self._context_manager = False
+        self._inside_finally: Optional[ast.Try] = None
+        self._scope: Optional[TrioScope] = None
 
     def visit_generic_with(self, node: Union[ast.With, ast.AsyncWith]):
         self.check_for_trio100(node)
 
-        outer = self._yield_is_error
-        if not self._context_manager and any(
-            is_trio_call(item, "open_nursery", *cancel_scope_names)
-            for item in (i.context_expr for i in node.items)
+        outer_yie = self._yield_is_error
+        outer_scope = self._scope
+
+        # Check for a `with trio.<scope_creater>`
+        trio_scope = next(
+            (
+                get_trio_scope(item, "open_nursery", *cancel_scope_names)
+                for item in (i.context_expr for i in node.items)
+            ),
+            None,
+        )
+        if trio_scope is not None:
+            if not self._context_manager:
+                self._yield_is_error = True
+
+            # Check for `with [...] as <varname>`
+            trio_scope.variable_name = next(
+                (
+                    item.optional_vars.id
+                    for item in node.items
+                    if isinstance(item.optional_vars, ast.Name)
+                ),
+                None,
+            )
+            self._scope = trio_scope
+        self.generic_visit(node)
+
+        # reset scope and yield_is_error
+        self._yield_is_error = outer_yie
+        self._scope = outer_scope
+
+    def visit_Assign(self, node: ast.Assign) -> None:
+        # checks for <scopename>.shield = [True/False]
+        if (
+            self._scope
+            and self._scope.variable_name is not None
+            and len(node.targets) == 1
+            and isinstance(node.targets[0], ast.Attribute)
+            and isinstance(node.targets[0].value, ast.Name)
+            and node.targets[0].value.id == self._scope.variable_name
+            and node.targets[0].attr == "shield"
+            and isinstance(node.value, ast.Constant)
         ):
-            self._yield_is_error = True
+            self._scope.shielded = node.value.value
+        self.generic_visit(node)
 
+    def visit_Await(self, node: ast.Await) -> None:
+        # if we're inside a finally, and not inside a context_manager, and we're either not in a scope, or in a scope that doesn't have both a timeout and shield
+        if (
+            self._inside_finally is not None
```

Comment:
I've just fixed my lazy writeup in #1, so it's my bad, but this check should also apply to `async for` and `async with` - i.e. all the `checkpoint_node_types` where the scheduler can kick in.